### DISCUSSION TITLE
Is there a way to manipulate a function's arguments and return types in a decorator?

### DISCUSSION THREAD

**Discussion by elonp** [microsoft/pyright#4405] [Q&A]
ParamSpecs and Concatenate allow decorators to add and remove arguments.  But that's a very limited scope of manipulation.  Is there a way to do more?

Crude example:
```python
def vectorize(f):
    def wf(*args):
        return [f(*t) for t in zip(*args)]

    return wf

@vectorize
def format_string_and_int(s: str, i: int) -> str:
    return f"{s},{i}"

if __name__ == "__main__":
    s = ["a", "b", "c"]
    i = [1, 2, 3]
    r = format_string_and_int(s, i)
    print(r)
```

output:
['a,1', 'b,2', 'c,3']

The type of `format_string_and_int` (after passing through the decorator) should be something like:
`Callable[[List[s], List[int]], List[str]]`

Is there a way for `vectorize` to declare that?

**Comment by erictraut** [comment]
I don't see any way to do this in the current type system. [PEP 646](https://peps.python.org/pep-0646/) gets you part of the way to a solution.

**Reply by elonp** [reply]
Alright.  Thank you.
